### PR TITLE
fix: buffer overrun when passing long command name from lua

### PR DESCRIPTION
**Pull request by romange**
Also, few additional changes that do not affect functionality.
1. make sure passed arguments to DispatchCommand are `\0` delimited during pipelining.
2. extend lua malloc hook to call precise functions - to help with cpu profiling.
3. reuse arguments buffer (save allocations) when calling Dragonfly command from lua scripts.

### CODE REVIEW

File: src/facade/dragonfly_connection.cc [hunk 1]
```diff
@@ -159,8 +159,9 @@ void Connection::PipelineMessage::SetArgs(const RespVec& args) {
     size_t s = buf.size();
     if (s)
       memcpy(next, buf.data(), s);
+    next[s] = '\0';
     this->args[i] = MutableSlice(next, s);
-    next += s;
+    next += (s + 1);
   }
```
Comment (dranikpg) [inline src/facade/dragonfly_connection.cc:165]: I don't understand. Do you rely now on the \0 to be present? But what about regular sync dispatch? What about StoredCmd (that is tightly packed)?

File: src/core/interpreter.cc [hunk 2]
```diff
@@ -805,9 +809,16 @@ int Interpreter::RedisGenericCommand(bool raise_error, bool async) {
...
+  redis_func_(
+      CallArgs{MutSliceSpan{args}, &buffer_, &translator, async, raise_error, &raise_error});
   cmd_depth_--;
 
+  // Shrink reusable buffer if it's too big.
+  if (buffer_.capacity() > 128) {
+    buffer_.clear();
+    buffer_.shrink_to_fit();
+  }
+
```
Comment (dranikpg) [inline src/core/interpreter.cc:821]: If the script does something like
```
while true:
  if not call: EXISTS key
    call HSET *lots of data*
```

you'll get a grow-shrink look and will actually have an allocation every single time 

Though its still better than allocating a buffer each time

**Review by dranikpg — APPROVED** [review]

File: src/facade/dragonfly_connection.cc (hunk 1 above)
Comment (romange) [reply src/facade/dragonfly_connection.cc:165]: I do not strictly rely on delimiter but I do want that the arguments have space between them to put '\0' delimiters when needed. The reason for this is what lua requires '\0' delimiters so I am planning to set them when passing the arguments to lua. out of scope of this PR.

**Comment by romange** [comment]
Added spacing in StoredCmd @dranikpg

**Review by dranikpg — APPROVED** [review]
What about direct invocation: No pipeline/no stored cmd, directly from redis_parser?

Also could it be that with increased storage overhead you decrease performance for edge cases by a few percent. We've cut out the command name to save a few bytes of storage for inlined strings, but now you've added that again 🙂

**Pull request merged**